src/ostrowski/rationals/basic.lean

```diff
@@ -0,0 +1,226 @@
+
+import data.nat.prime
+import data.rat.basic
+import data.real.basic
+import analysis.special_functions.pow
+import data.real.cau_seq
+import data.padics
+
+open is_absolute_value
+noncomputable theory
+
+def trivial_abs : ℚ → ℝ := λ a,
+    if a = 0 then 0
+    else 1
+
+instance : is_absolute_value trivial_abs :=
+{ abv_nonneg := λ x, begin rw [trivial_abs], dsimp, split_ifs, exact le_refl 0, exact zero_le_one, end,
+  abv_eq_zero := λ x, begin rw [trivial_abs], dsimp, split, contrapose!, intro, rw if_neg ᾰ, norm_num, intro, rw if_pos ᾰ, end,
+  abv_add := λ x y, begin rw [trivial_abs], dsimp, split_ifs, any_goals { linarith,}, exfalso, rw [h_1,h_2] at h, simpa using h, end,
+  abv_mul := λ x y, begin rw [trivial_abs], dsimp, split_ifs, any_goals {norm_num}, finish, finish, finish, finish, end }
+
+def padic_norm_ℝ (p: ℕ) := λ r: ℚ, (padic_norm p r: ℝ)
+
+instance abv_lift (abv: ℚ → ℚ) [habv: is_absolute_value abv]:
+  is_absolute_value (λ r: ℚ, (abv r: ℝ)) :=
+{ abv_nonneg := by { norm_cast, exact abv_nonneg abv, },
+  abv_eq_zero := by { norm_cast, intro x, exact abv_eq_zero abv, },
+  abv_add := by { norm_cast, exact abv_add abv, },
+  abv_mul := by { norm_cast, exact abv_mul abv, } }
+
+instance padic_is_abv {p: ℕ} [hp: nat.prime p] : is_absolute_value (padic_norm_ℝ p) :=
+  @abv_lift (padic_norm p) (@padic_norm.is_absolute_value p hp)
+
+lemma trivial_abs_is_one_iff_nonzero_arg (a: ℚ) : (a ≠ 0) ↔ (trivial_abs a = 1) :=
+begin
+    split,
+    intro h₁,
+    unfold trivial_abs,
+    by_cases (a = 0),
+    exfalso,
+    exact h₁ h,
+    rw if_neg h,
+    contrapose!,
+    intro h,
+    unfold trivial_abs,
+    rw if_pos h,
+    finish
+end
+
+lemma rat_mk_eq_div (x_num : ℤ) (x_denom : ℕ) (x_pos : 0 < x_denom) (x_cop : nat.coprime (int.nat_abs x_num) x_denom)
+: ({num := x_num, denom := x_denom, pos := x_pos, cop := x_cop}:ℚ) = (↑ x_num)/(↑ x_denom) := begin
+norm_cast,
+rw rat.mk,
+rw rat.mk_nat,
+rw dif_neg,
+rw rat.mk_pnat,
+{
+  simp [x_cop.gcd_eq_one],
+},
+exact ne_of_gt x_pos,
+end
+
+def hom_of_abv {α} [linear_ordered_field α] {β} [ring β] [nontrivial β]
```

Comment:
hm why nontrivial is needed, is that a condition imposed by hom?

Reply:
same for ring, is it needed? wouldn't semiring be enough?

Reply:
Both are required by `is_absolute_value.abv_one`

Reply:
This could be generalized, but I do not really see any point in letting `β` be trivial as long as this lemma does not go into mathblib.